pkg/daemon/update.go

```diff
@@ -187,10 +192,28 @@ func (dn *Daemon) update(oldConfig, newConfig *mcfgv1.MachineConfig) error {
 		return err
 	}
 
+	defer func() {
+		if retErr != nil {
+			if err := dn.updateFiles(newConfig, oldConfig); err != nil {
+				retErr = errors.Wrapf(retErr, "error rolling back files writes %v", err)
+				return
+			}
+		}
+	}()
+
 	if err := dn.updateSSHKeys(newConfig.Spec.Config.Passwd.Users); err != nil {
 		return err
 	}
 
+	defer func() {
+		if retErr != nil {
+			if err := dn.updateSSHKeys(oldConfig.Spec.Config.Passwd.Users); err != nil {
```

Comment:
Very cool!

Reply:
This functionality looks like it's working fine, added an error in the MCD simulating a...error and it goes back and rewrite files from oldConfig just fine (I expect to work the same for ssh keys)

```
I0314 16:51:27.336603    6697 update.go:187] Checking reconcilable for config worker-2231f5479a086534c246ec862625a3f1 to worker-41c30245368d268544394d45a2e70598
I0314 16:51:27.336628    6697 update.go:245] Checking if configs are reconcilable
I0314 16:51:27.338097    6697 update.go:385] Updating files
I0314 16:51:27.338118    6697 update.go:586] Writing file "/etc/sysconfig/crio-network"
I0314 16:51:27.344087    6697 update.go:586] Writing file "/var/lib/kubelet/config.json"
I0314 16:51:27.347939    6697 update.go:586] Writing file "/etc/kubernetes/ca.crt"
I0314 16:51:27.356594    6697 update.go:586] Writing file "/etc/sysctl.d/forward.conf"
I0314 16:51:27.359890    6697 update.go:586] Writing file "/etc/kubernetes/kubelet-plugins/volume/exec/.dummy"
I0314 16:51:27.362261    6697 update.go:586] Writing file "/etc/containers/registries.conf"
I0314 16:51:27.365901    6697 update.go:586] Writing file "/etc/containers/storage.conf"
I0314 16:51:27.369021    6697 update.go:586] Writing file "/etc/crio/crio.conf"
I0314 16:51:27.372757    6697 update.go:586] Writing file "/etc/kubernetes/kubelet.conf"
I0314 16:51:27.376479    6697 update.go:586] Writing file "/etc/chrony.conf"
I0314 16:51:27.380001    6697 update.go:523] Writing systemd unit "kubelet.service"
I0314 16:51:27.383001    6697 update.go:558] Enabling systemd unit "kubelet.service"
I0314 16:51:27.383052    6697 update.go:476] /etc/systemd/system/multi-user.target.wants/kubelet.service already exists. Not making a new symlink
I0314 16:51:27.383063    6697 update.go:404] Deleting stale data
I0314 16:51:27.383084    6697 update.go:648] Writing SSHKeys at "/home/core/.ssh/authorized_keys"
I0314 16:51:27.385047    6697 daemon.go:1009] Current and target osImageURL have matching digest "sha256:c09f455cc09673a1a13ae7b54cc4348cda0411e06dfa79ecd0130b35d62e8670"
I0314 16:51:27.385071    6697 update.go:648] Writing SSHKeys at "/home/core/.ssh/authorized_keys"
I0314 16:51:27.386892    6697 update.go:385] Updating files
I0314 16:51:27.386910    6697 update.go:586] Writing file "/etc/sysconfig/crio-network"
I0314 16:51:27.388912    6697 update.go:586] Writing file "/var/lib/kubelet/config.json"
I0314 16:51:27.393087    6697 update.go:586] Writing file "/etc/kubernetes/ca.crt"
I0314 16:51:27.396004    6697 update.go:586] Writing file "/etc/sysctl.d/forward.conf"
I0314 16:51:27.398847    6697 update.go:586] Writing file "/etc/kubernetes/kubelet-plugins/volume/exec/.dummy"
I0314 16:51:27.400039    6697 update.go:586] Writing file "/etc/containers/registries.conf"
I0314 16:51:27.402039    6697 update.go:586] Writing file "/etc/containers/storage.conf"
I0314 16:51:27.407845    6697 update.go:586] Writing file "/etc/crio/crio.conf"
I0314 16:51:27.413978    6697 update.go:586] Writing file "/etc/kubernetes/kubelet.conf"
I0314 16:51:27.417781    6697 update.go:523] Writing systemd unit "kubelet.service"
I0314 16:51:27.420848    6697 update.go:558] Enabling systemd unit "kubelet.service"
I0314 16:51:27.420897    6697 update.go:476] /etc/systemd/system/multi-user.target.wants/kubelet.service already exists. Not making a new symlink
I0314 16:51:27.420908    6697 update.go:404] Deleting stale data
I0314 16:51:27.421027    6697 daemon.go:396] Unable to apply update: test
E0314 16:51:27.421053    6697 writer.go:98] marking degraded due to: test
```

Reply:
so to understand, this tried to do the first config, for whatever reason it can't, so it goes back to writing the previous config that didnt have the chrony.conf & annotates as degraded?

Reply:
yep, and keeps trying to apply the new config as expected by a "controller" which keeps retrying (even if in this case, the obvious fix would be to remove the MC, create a new correct one, and reconcile, this was just a smoke test)

Reply:
so going forward to recover from a bad mc, it's just now make a new mc or make a new mc & remove degraded annotation? (i think we should probably write this down for users?)

Reply:
recover from a bad mc is just 1) delete the bad MC first, 2) create a new MC with something "good" which applies

Reply:
I'm wondering if we should, in the future, react to deletions and just undo the whole generated machineconfigs altogether

Reply:
I do actually believe that deleting an MC is triggering a regeneration of the generated MC for the pool (according to the render_controller code)